src/Turtle/Line.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DeriveDataTypeable         #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Turtle.Line
+  ( Line(lineToText)
+  , textToLines
+  , linesToText
+  , textToLine
+  , unsafeTextToLine
+  , NewlineForbidden(..)
+  ) where
+
+import Data.Text (Text)
+import qualified Data.Text as Text
+#if __GLASGOW_HASKELL__ >= 708
+import Data.Coerce
+#endif
+import Data.String
+#if __GLASGOW_HASKELL__ >= 710
+#else
+import Data.Monoid
+#endif
+import Data.Typeable
+import Control.Exception
+
+data NewlineForbidden = NewlineForbidden
+  deriving (Show, Typeable)
+
+instance Exception NewlineForbidden
+
+-- | A line of text (does not contain newlines).
+newtype Line = Line { lineToText :: Text }
+  deriving (Eq, Ord, Show, Monoid)
+
+instance IsString Line where
+  fromString = textToLine . fromString
+
+textToLines :: Text -> [Line]
+textToLines = mapTextToLine . Text.lines
+  where
+    mapTextToLine :: [Text] -> [Line]
+    mapTextToLine =
+#if __GLASGOW_HASKELL__ >= 708
+      coerce
+#else
+      map unsafeTextToLine
+#endif
+
+linesToText :: [Line] -> Text
+linesToText = Text.unlines . mapLineToText
+  where
+    mapLineToText :: [Line] -> [Text]
+    mapLineToText =
+#if __GLASGOW_HASKELL__ >= 708
+      coerce
+#else
+      map lineToText
+#endif
+
+-- | Precondition (checked): the argument does not contain newlines.
+-- Throws: `NewlineForbidden`.
+-- You probably want `textToLines` instead.
+textToLine :: Text -> Line
```

Comment:
I feel like this function should return `Maybe` and only the `fromString` instance should `throw`